lightning/src/ln/channelmanager.rs

```diff
@@ -3321,6 +3383,102 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+				.expect("RNG Generated Duplicate PaymentHash"))
+	}
+
+	/// Gets a [`PaymentSecret`] for a given [`PaymentHash`], for which the payment preimage is
+	/// stored external to LDK.
+	///
+	/// A [`PaymentReceived`] event will only be generated if the [`PaymentSecret`] matches a
+	/// payment secret fetched via this method or [`create_inbound_payment`], and which is at least
+	/// the `min_value_msat` provided here, if one is provided.
+	///
+	/// The [`PaymentHash`] (and corresponding [`PaymentPreimage`]) must be globally unique. This
+	/// method may return an Err if another payment with the same payment_hash is still pending.
+	///
+	/// `user_payment_id` will be provided back in [`PaymentReceived::user_payment_id`] events to
+	/// allow tracking of which events correspond with which calls to this and
+	/// [`create_inbound_payment`]. `user_payment_id` has no meaning inside of LDK, it is simply
+	/// copied to events and otherwise ignored. It may be used to correlate PaymentReceived events
+	/// with invoice metadata stored elsewhere.
+	///
+	/// `min_value_msat` should be set if the invoice being generated contains a value. Any payment
+	/// received for the returned [`PaymentHash`] will be required to be at least `min_value_msat`
+	/// before a [`PaymentReceived`] event will be generated, ensuring that we do not provide the
+	/// sender "proof-of-payment" unless they have paid the required amount.
+	///
+	/// `invoice_expiry_delta_secs` describes the number of seconds that the invoice is valid for
+	/// in excess of the current time. This should roughly match the expiry time set in the invoice.
+	/// After this many seconds, we will remove the inbound payment, resulting in any attempts to
+	/// pay the invoice failing. The BOLT spec suggests 7,200 secs as a default validity time for
+	/// invoices when no timeout is set.
+	///
+	/// Note that we use block header time to time-out pending inbound payments (with some margin
+	/// to compensate for the inaccuracy of block header timestamps). Thus, in practice we will
+	/// accept a payment and generate a [`PaymentReceived`] event for some time after the expiry.
+	/// If you need exact expiry semantics, you should enforce them upon receipt of
+	/// [`PaymentReceived`].
+	///
+	/// May panic if `invoice_expiry_delta_secs` is greater than one year.
+	///
+	/// Note that invoices generated for inbound payments should have their `min_final_cltv_expiry`
+	/// set to at least [`MIN_FINAL_CLTV_EXPIRY`].
+	///
+	/// [`create_inbound_payment`]: Self::create_inbound_payment
+	/// [`PaymentReceived`]: events::Event::PaymentReceived
+	/// [`PaymentReceived::user_payment_id`]: events::Event::PaymentReceived::user_payment_id
+	pub fn create_inbound_payment_for_hash(&self, payment_hash: PaymentHash, min_value_msat: Option<u64>, invoice_expiry_delta_secs: u32, user_payment_id: u64) -> Result<PaymentSecret, APIError> {
```

Comment:
Good with the API, only concern is a limit `pending_inbound_payments`(`MAX_PENDING_PAYMENTS`=10000). Otherwise if your payment logic generates invoices on third-party demand and call `create_inbound_payment_for_hash` in consequence you have a DoS vector ? Even if I don't know how much we should care about unsafe use of the API this one sounds pretty easy to mitigate...

Reply:
Hmmm, that's definitely a good question. I don't really know that that's a requirement at our layer - certainly some users (especially those on a server providing access to inbound payments via an API) may want a huge limit, and others a very small one. Presumably you also want rate-limiting by the client, not just a strict limit. We can certainly mention it in the docs, is that sufficient?

Reply:
Mentioned in #905.